bench.py

```diff
@@ -0,0 +1,162 @@
+logging.basicConfig(
+        format="%(asctime)s - %(levelname)s - %(name)s -   %(message)s",
+        datefmt="%m/%d/%Y %H:%M:%S",
+        level=logging.INFO,
+)
+
+# -----------------------------------------------------------------------------
+import os
+if int(os.environ.get('USE_LIGHTNING', 0)):
+    logging.info("USING LIGHTNING!!")
+    import pytorch_lightning as pl
+else:
+    import mingpt.fake_lightning as pl
+    logging.info("using our humble trainer")
+# -----------------------------------------------------------------------------
+
+class CharDataset(Dataset):
+    """
+    e.g. Text8 dataset is often used: http://mattmahoney.net/dc/textdata.html
+    Vocabulary is lowercase English characters and space for total of 27.
+    Training data: First 90M characters.
+    Validation data: First 5M characters out of the last 10M characters.
+    Testing data: Last 5M characters.
+    """
+
+    def __init__(self, data_path, block_size, split, override_vocab=None):
+
+        # load the data and crop it appropriately
+        with open(data_path, 'r') as f:
+            data = f.read()
+
+        crop = {
+            'train': (0, 90000000),
+            'val': (90000000, 95000000),
+            'test': (95000000, 100000000),
+        }[split]
+        data = data[crop[0]:crop[1]]
+
+        # build a vocabulary from data or inherit it
+        vocab = sorted(list(set(data))) if override_vocab is None else override_vocab
+        data_size, vocab_size = len(data), len(vocab)
+        logging.info('data of crop %s has %d characters, vocab of size %d.' % (str(crop), data_size, vocab_size))
+
+        self.stoi = { ch:i for i,ch in enumerate(vocab) }
+        self.itos = { i:ch for i,ch in enumerate(vocab) }
+        self.block_size = block_size
+        self.vocab_size = vocab_size
+        self.data = data
+        self.vocab = vocab
+
+    def __len__(self):
+        return math.ceil(len(self.data) / self.block_size)
+
+    def __getitem__(self, idx):
+        i = np.random.randint(0, len(self.data) - (self.block_size + 1))
+        chunk = self.data[i:i+self.block_size+1]
+        dix = torch.tensor([self.stoi[s] for s in chunk], dtype=torch.long)
+        return dix[:-1], dix[1:]
+
+class CharDataModule(pl.LightningDataModule):
+
+    def __init__(self, batch_size=64, block_size=128, pin_memory=0, num_workers=0):
+        super().__init__()
+        self.batch_size = batch_size
+        self.block_size = block_size
+        self.num_workers = num_workers
+        self.pin_memory = pin_memory
+
+        self.train_dataset = CharDataset('text8', self.block_size, 'train')
+
+    def prepare_data(self): # called only on 1 GPU/machine
+        pass # could technically download text8 here...
+
+    def setup(self, stage): # called for every GPU/machine
+        if stage == 'train' or stage == 'fit':
+            pass # nothing to do, the train_dataset is initialized in the constructor
+        elif stage == 'val':
```

Comment:
stage here is whether the setup is happening on fit or test...

BUT... to your point, do you think it might make more sense to change the stage to 'train', 'val', 'test'? the thing is that train, val are usually handled together (ie: split train into train/val and have a separate test set)

Reply:
Sorry I'm pretty sure I totally misunderstood what a "stage" is and thought it referred to splits.